// plivo/dist/rest/axios.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

exports.Axios = Axios;

var _exceptions = require('../utils/exceptions');

var Exceptions = _interopRequireWildcard(_exceptions);

var _lodash = require('lodash');

var _ = _interopRequireWildcard(_lodash);

var _axios = require('axios');

var _axios2 = _interopRequireDefault(_axios);

var _querystring = require('querystring');

var _querystring2 = _interopRequireDefault(_querystring);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

var HttpsProxyAgent = require('https-proxy-agent');

function Axios(config) {
  var auth = 'Basic ' + new Buffer(config.authId + ':' + config.authToken).toString('base64');

  var headers = {
    Authorization: auth,
    'User-Agent': config.userAgent,
    'Content-Type': 'application/json'
  };

  var retryWrapper = function retryWrapper(axios, options) {
    var max_time = options.retryTime;
    var counter = 0;
    axios.interceptors.response.use(null, function (error) {
      var config = error.config;
      if (counter < max_time && error.response.status >= 500) {
        counter++;
        config.url = options.urls[counter] + options.authId + '/' + options.action;
        return new Promise(function (resolve) {
          resolve(axios(config));
        });
      }
      return Promise.reject(error);
    });
  };

  return function (method, action, params) {
    var configuration = config;
    // Add support fot multipart requests.
    if (typeof params != 'undefined' && params.multipart == true) {
      return new Promise(function (resolve, reject) {
        delete params.multipart;

        var FormData = require('form-data');
        var multipartParams = new FormData();

        for (var key in params) {
          if (key != 'file') {
            multipartParams.append(key, params[key]);
          } else {
            multipartParams.append(key, require('fs').createReadStream(params[key]));
          }
        }

        var headers = multipartParams.getHeaders();

        var config = {
          method: method,
          url: configuration.url + '/' + action,
          headers: {
            'Authorization': 'Basic ' + new Buffer(configuration.authId + ':' + configuration.authToken).toString('base64'),
            'User-Agent': configuration.userAgent,
            'content-type': headers['content-type']
          },
          data: multipartParams
        };

        (0, _axios2.default)(config).then(function (response) {
          var exceptionClass = {
            400: Exceptions.InvalidRequestError,
            401: Exceptions.AuthenticationError,
            404: Exceptions.ResourceNotFoundError,
            405: Exceptions.InvalidRequestError,
            500: Exceptions.ServerError
          }[response.status] || Error;

          if (!_.inRange(response.status, 200, 300)) {
            var body = response.data;
            if ((typeof body === 'undefined' ? 'undefined' : _typeof(body)) === 'object') {
              reject(new exceptionClass(JSON.stringify(body)));
            } else {
              reject(new exceptionClass(body));
            }
          }
          resolve({
            response: response,
            body: response.data
          });
        }).catch(function (error) {
          var exceptionClass = {
            400: Exceptions.InvalidRequestError,
            401: Exceptions.AuthenticationError,
            404: Exceptions.ResourceNotFoundError,
            405: Exceptions.InvalidRequestError,
            500: Exceptions.ServerError
          }[error.response.status] || Error;
          if (!_.inRange(error.response.status, 200, 300)) {
            var body = error.response.data;
            if ((typeof body === 'undefined' ? 'undefined' : _typeof(body)) === 'object') {
              reject(new exceptionClass(error));
            } else {
              if (error.response.status >= 500) {
                reject(new Exceptions.ServerError(error));
              }
              reject(new exceptionClass(error));
            }
          }
          reject(error.stack + '\n' + JSON.stringify(error.response.data));
        });
      });
    }

    if (typeof params != 'undefined' && typeof params.file != 'undefined') {
      var files = [];
      if (Array.isArray(params.file)) {
        for (var index = 0; index < params.file.length; index++) {
          files[index] = require('fs').createReadStream(params.file[index]);
        }
      } else {
        files[0] = require('fs').createReadStream(params.file);
      }
      params.file = files;
    }

    var options = {
      url: config.url + '/' + action,
      method: method,
      data: params || '',
      headers: headers,
      json: true
    };
    var apiVoiceUris = ['https://voice.plivo.com/v1/Account/', 'https://voice-usw1.plivo.com/v1/Account/', 'https://voice-use1.plivo.com/v1/Account/'];
    var isVoiceReq = false;
    if (params) {
      if (params.hasOwnProperty('is_call_insights_request')) {
        options.url = params.call_insights_base_url + params.call_insights_request_path;
        delete params.is_call_insights_request;
        delete params.call_insights_base_url;
        delete params.call_insights_request_path;
        delete options.data;
        options.json = params;
      } else if (params.hasOwnProperty('is_voice_request')) {
        options.url = apiVoiceUris[0] + config.authId + '/' + action;
        delete params.is_voice_request;
        isVoiceReq = true;
      } else if (params.hasOwnProperty('override_url')) {
        // currently used by Lookup API but is generic enough to be used
        // by any product in future.
        options.url = params.override_url;
        delete params.override_url;
      }
    }

    if (method === 'GET' && options.data !== '') {
      var query = '?' + _querystring2.default.stringify(params);
      options.url += query;
    }

    if (typeof config.proxy !== 'undefined') {
      options.httpsAgent = new HttpsProxyAgent(config.proxy);
    }

    if (typeof config.timeout !== 'undefined') {
      options.timeout = config.timeout;
    }

    return new Promise(function (resolve, reject) {
      if (isVoiceReq) {
        retryWrapper(_axios2.default, { retryTime: 2, urls: apiVoiceUris, authId: config.authId, action: action });
        options.url = apiVoiceUris[0] + config.authId + '/' + action;
        (0, _axios2.default)(options).then(function (response) {
          var exceptionClass = {
            400: Exceptions.InvalidRequestError,
            401: Exceptions.AuthenticationError,
            404: Exceptions.ResourceNotFoundError,
            405: Exceptions.InvalidRequestError,
            500: Exceptions.ServerError
          }[response.status] || Error;

          if (!_.inRange(response.status, 200, 300)) {
            var body = response.data;
            if ((typeof body === 'undefined' ? 'undefined' : _typeof(body)) === 'object') {
              reject(new exceptionClass(JSON.stringify(body)));
            } else {
              reject(new exceptionClass(body));
            }
          }
          resolve({
            response: response,
            body: response.data
          });
        }).catch(function (error) {
          var exceptionClass = {
            400: Exceptions.InvalidRequestError,
            401: Exceptions.AuthenticationError,
            404: Exceptions.ResourceNotFoundError,
            405: Exceptions.InvalidRequestError,
            500: Exceptions.ServerError
          }[error.response.status] || Error;
          if (!_.inRange(error.response.status, 200, 300)) {
            var body = error.response.data;
            if ((typeof body === 'undefined' ? 'undefined' : _typeof(body)) === 'object') {
              reject(new exceptionClass(error));
            } else {
              if (error.response.status >= 500) {
                reject(new Exceptions.ServerError(error));
              }
              reject(new exceptionClass(error));
            }
          }
          reject(error.stack + '\n' + JSON.stringify(error.response.data));
        });
      } else {
        (0, _axios2.default)(options).then(function (response) {
          var exceptionClass = {
            400: Exceptions.InvalidRequestError,
            401: Exceptions.AuthenticationError,
            404: Exceptions.ResourceNotFoundError,
            405: Exceptions.InvalidRequestError,
            500: Exceptions.ServerError
          }[response.status] || Error;

          if (!_.inRange(response.status, 200, 300)) {
            var body = response.data;
            if ((typeof body === 'undefined' ? 'undefined' : _typeof(body)) === 'object') {
              reject(new exceptionClass(JSON.stringify(body)));
            } else {
              reject(new exceptionClass(body));
            }
          } else {
            var _body2 = response.data;
            var isObj = (typeof _body === 'undefined' ? 'undefined' : _typeof(_body)) === 'object' && _body !== null && !(_body instanceof Array) && !(_body instanceof Date);
            if (isObj) {
              _body['statusCode'] = response.status;
            }
            resolve({
              response: response,
              body: _body2
            });
          }
        }).catch(function (error) {
          var exceptionClass = {
            400: Exceptions.InvalidRequestError,
            401: Exceptions.AuthenticationError,
            404: Exceptions.ResourceNotFoundError,
            405: Exceptions.InvalidRequestError,
            500: Exceptions.ServerError
          }[error.response.status] || Error;
          if (!_.inRange(error.response.status, 200, 300)) {
            var body = error.response.data;
            if ((typeof body === 'undefined' ? 'undefined' : _typeof(body)) === 'object') {
              reject(new exceptionClass(error));
            } else {
              if (error.response.status >= 500) {
                reject(new Exceptions.ServerError(error));
              }
              reject(new exceptionClass(error));
            }
          }
          reject(error.stack + '\n' + JSON.stringify(error.response.data));
        });
      }
    });
  };
}
